featuregates/controller/pkg/feature/feature_controller.go

```diff
@@ -0,0 +1,242 @@
+	}
+	return nil
+}
+
+// reconcileDeletedFeature reconciles feature that has been deleted
+func reconcileDeletedFeature(ctx context.Context, c client.Client, featureName string) error {
+	// Check if feature part of any FeatureGate spec and update status in Feature result to Invalid
+	featureGate, found, err := util.GetFeatureGateForFeature(ctx, c, featureName)
+	if err != nil {
+		return err
+	}
+	if found {
+		featureGate.Status.Results = computeFeatureGateStatusResults(featureGate.Status, corev1alpha2.Result{
+			Name:    featureName,
+			Status:  constants.InvalidFeatureReferenceStatus,
+			Message: "Feature doesn't exist in cluster",
+		}, true)
+		if err := c.Status().Update(ctx, featureGate); err != nil {
+			return fmt.Errorf("couldn't update %s FeatureGate status :%w", featureGate.Name, err)
+		}
+	}
+	return nil
+}
+
+// reconcileFeatureNotInFeatureGateSpec reconciles feature that is not found in any FeatureGate spec
+func reconcileFeatureNotInFeatureGateSpec(ctx context.Context, c client.Client, feature *corev1alpha2.Feature) error {
+	// If feature is not part of any FeatureGate spec, check for its existence in the FeatureGate status.
+	// If found in the FeatureGate status, remove its entry from status and update the feature status to deactivated
+	featureGate, found, err := util.GetFeatureGateWithFeatureInStatus(ctx, c, feature.Name)
+	if err != nil {
+		return err
+	}
+	if found {
+		// Remove feature from FeatureGate status
+		featureGate.Status.Results = computeFeatureGateStatusResults(featureGate.Status, corev1alpha2.Result{
+			Name: feature.Name,
+		}, false)
+		if err := c.Status().Update(ctx, featureGate); err != nil {
+			return fmt.Errorf("couldn't update %s FeatureGate status :%w", featureGate.Name, err)
+		}
+	}
+	// Update Feature status to set feature as deactivated
+	feature.Status.Activated = false
+	if err := c.Update(ctx, feature); err != nil {
+		return fmt.Errorf("couldn't update %s Feature status :%w", feature.Name, err)
+	}
+	return nil
+}
+
+// applyPolicyToComputeFeatureResultAndActivation applies stability level policy and returns feature result for FeatureGate status and feature activate status
+func applyPolicyToComputeFeatureResultAndActivation(policy corev1alpha2.Policy, featureRef corev1alpha2.FeatureReference, stabilityLevel string) (corev1alpha2.Result, bool) {
+	activated := policy.DefaultActivation
+	result := corev1alpha2.Result{Name: featureRef.Name}
+	// Check for immutability and change in intent of feature status
+	if policy.Immutable && !featureRef.SkipStabilityPolicyValidation && policy.DefaultActivation != featureRef.Activate {
+		result.Status = constants.InvalidFeatureReferenceStatus
+		result.Message = "Feature couldn't be toggled because it is immutable"
+	} else if policy.VoidsWarranty && !featureRef.SkipStabilityPolicyValidation {
+		result.Status = constants.InvalidFeatureReferenceStatus
+		result.Message = "Toggling this feature voids the warranty"
```

Comment:
**nitpick**: Could this message be clearer? (E.g., "The stability level of this feature indicates that it should not be activated in production environments. To activate the feature, you must agree to permanently void all support guarantees for this environment by setting PermanentlyVoidAllSupportGuaranteesForThisEnvironment = true.")